### PR TITLE
feat: push down pruning for delete

### LINKED ISSUE #12933: feat: push down pruning for delete
**Summary**

Like read.

https://github.com/datafuselabs/databend/blob/main/src/query/storages/fuse/src/operations/read_partitions.rs#L82-L120

https://github.com/datafuselabs/databend/blob/main/src/query/storages/fuse/src/operations/read_data.rs#L147-L190

### PR DESCRIPTION
**Pull request by SkyFan2002**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Summary about this PR

- Closes #12933

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/12936)

### CODE REVIEW

File: src/query/service/src/interpreters/interpreter_delete.rs [hunk 1]
```diff
@@ -234,10 +235,15 @@ impl Interpreter for DeleteInterpreter {
             )
             .await?
         {
+            let mut segments = Vec::with_capacity(snapshot.segments.len());
```
Comment (zhyass) [inline src/query/service/src/interpreters/interpreter_delete.rs:238]: If the number of segments is too small, there will be a problem of  the data skew.

File: src/query/service/src/pipelines/pipeline_builder.rs [hunk 2]
```diff
@@ -795,9 +801,53 @@ impl PipelineBuilder {
...
+                            .await
+                    })?;
+                info!(
+                    "delete pruning done, number of whole block deletion detected in pruning phase: {}",
+                    info.num_whole_block_mutation
+                );
+                ctx.set_partitions(partitions)?;
+                Ok(())
+            });
+        }
```
Comment (zhyass) [inline src/query/service/src/pipelines/pipeline_builder.rs:847]:
```suggestion
        } else {
            ctx.set_partitions(delete.parts.clone())?;
        }
```

File: src/query/storages/fuse/src/operations/delete.rs [hunk 3]
```diff
@@ -127,26 +126,7 @@ impl FuseTable {
...
-            .await?;
-        info!(
-            "fast delete done, number of whole block deletion detected in pruning phase: {}",
-            info.num_whole_block_mutation
-        );
-        if partitions.is_empty() {
-            return Ok(None);
-        }
-        Ok(Some((partitions, snapshot.clone())))
+        Ok(Some(snapshot.clone()))
```
Comment (zhyass) [inline src/query/storages/fuse/src/operations/delete.rs:129]: remove parts in `add_deletion_source`

**Comment by SkyFan2002** [comment]
Tested by `test_conflict` script in cluster mode, local disk.
Log:https://github.com/SkyFan2002/databend/blob/refactor_long_run/tests/test_conflict/test.log

File: src/query/service/src/pipelines/pipeline_builder.rs [hunk 4]
```diff
@@ -795,13 +801,56 @@ impl PipelineBuilder {
...
+                });
+            }
+            let prune_ctx = MutationBlockPruningContext {
+                segment_locations,
+                block_count: None,
+            };
+            self.main_pipeline.set_on_init(move || {
+                let ctx_clone = ctx.clone();
+                let (partitions, info) =
+                    Runtime::with_worker_threads(1, None)?.block_on(async move {
```
Comment (zhyass) [inline src/query/service/src/pipelines/pipeline_builder.rs:826]:
```suggestion
                    Runtime::with_worker_threads(2, None)?.block_on(async move {
```

**Review by zhyass — APPROVED** [review]

**Review by dantengsky — APPROVED** [review]

**Pull request merged**